### ISSUE TITLE
Problems with Full Third-Party Cookie Blocking

### ISSUE THREAD

**Issue by somuelle-tmt** [bigbluebutton/bigbluebutton#9998]
**Describe the bug**
Embedding BBB inside an Iframe in a Browser with Full Third-Party Cookie Blocking activated leads to a 401 forbidden error.

**To Reproduce**
1. Use a Browser like Safari or configure Full Third-Party Cookie Blocking in a browser like Firefox
2. Use API to create Meeting an authenticate user
3. Embed BBB inside an Iframe

**Expected behavior**
Good question. Maybe an explanation like "Your browser blocks cookies. To enter this room allow cookies for bbb.example.com or click here to try opening in a new tab"?
I don't know how easy it is to detect Full Third-Party Cookie Blocking

**Actual behavior**
401 forbidden error

**Screenshots**
![Fehlermeldung](https://user-images.githubusercontent.com/11993445/86766989-98fcb580-c04b-11ea-838b-1073b67ac890.png)

**Desktop (please complete the following information):**
 - OS: every
 - Browser: Safari default, other Browsers can be configured
 - Version: every

**Smartphone (please complete the following information):**
 - Device: every
 - OS: every
 - Browser Safari default, other Browsers can be configured
 - Version: every

**Additional context**
Because you can find error #1958 at Google with the terms "BBB iframe safari" I would like to link it. But since it's about the old Flash client, the issue will probably not be reopened.

**Comment by basisbit** [comment]
Your website blocks cookies for third party websites and you did not add your bbb domain to your allowlist. Thus, your webbrowser tried to open the meeting website without any authentication cookies. In such a case, the HTTP 401 error page is the correct behaviour. It is best practices to not show possible attackers what exactly failed any security checks. In the bbb logs, you'll see more details about why the HTTP 401 was returned.

**Comment by somuelle-tmt** [comment]
I don't think that a website can block 3rd party cookies, this is a behavior of the browser. Which allowlist do you mean?
In the default settings of Firefox, Edge and Chrome everything works fine, and even if you disable "Full Third-Party Cookie Blocking" in Safari it works.

A failed login can hardly be called an "attack". According to the logs, the error comes because there are no cookies. Why this is the case, the BBB can not know.

**Comment by snadal** [comment]
I’m not sure if BBB is intended to be used inside frames, but for 3rd party cookies to work inside an iframe, it should use Storage Access API. 

See https://webkit.org/blog/8124/introducing-storage-access-api/ for a brief introduction.

This would allow BBB to be embedded inside iframes in external websites.

**Comment by somuelle-tmt** [comment]
@snadal Good point! Since BBB in the current version documents parameters especially for iframes (which are used in our LMS) I would say yes: https://docs.bigbluebutton.org/2.2/customize.html#external-parameters

Storage Access API sounds like a good solution! According to [this site](https://www.chromestatus.com/feature/5612590694662144) it's already shipped in Firefox und was [rolled out in Canary and Dev channels of Edge yesterday](https://blogs.windows.com/msedgedev/2020/07/08/introducing-storage-access-api/).

**Comment by f1pratyush** [comment]
I encountered this problem too and after getting [some help from StackOverflow](https://stackoverflow.com/a/63234717/8602699), I could narrow down the issue. The issue is that the BBB Client is trying to set up a cookie and is unable to do so because it's in an iframe and Chrome has updated it's cookie policies.

Obviously, if you are building an app you can't expect all users to apply the fix mentioned in the link above. What needs to be done is to configure BBB to set the cookie with properties "SameSite=None" and "Secure".

Right now, I am combing through the BBB Code to find where exactly is the cookie method called and make that change.

Hopefully, BBB fixes this in a later update.

**Comment by matiasilva** [comment]
@f1pratyush If you are interested, please submit a PR.

**Comment by pedrobmarin** [comment]
We did manage to fix the shared notes' Etherpad server recently to always set it's cookies to `SameSite=None` and `secure` (when behind a secured connection) because it always runs in an iframe https://github.com/bigbluebutton/etherpad-lite/commit/aafba502a4d19db8fd255c4c2fee9b3c8812df28.

Although might look like you could embed the whole bigbluebutton client in an iframe, that's not recommended. But in case you have no other alternative besides running it inside an iframe, try checking out how to configure that using nginx' `proxy_cookie_path`.

**Comment by athanasius-kircher** [comment]
I think there is a different option, according to https://groups.google.com/g/bigbluebutton-dev/c/CP0trA-AKR0/m/UZEhBFtFAwAJ, BigBlueButton has an option: `allowRequestsWithoutSession`. That seems to eliminate the need to have a session and hence cookies.

As far as I can see that may be the better option for these integrations via iframe, as bbb does not require the cookie anymore. Hence you would not have to care about third party cookie handling in different browsers and continuous behaviour changes in browsers.

In `/usr/share/bbb-web/WEB-INF/classes/bigbluebutton.properties` I set `allowRequestsWithoutSession=true`. Restarted BBB and the iframe inclusion worked without the cookie.

**Implications**
I see that now the sessionToken (in the join url) is not bound to a Serversession identified by the cookie and hence the owner of the session link containing the token may join the session. That is indeed a weaker security measurement. 

**My questions to the BBB-Team**
1. Is that an intended configuration option?
2. Are there any other side effects when using this option?
3. Would you assess the degree of security high enough?
4. If 1 - 3 are are evaluated positive, I would recommend a doc extension.

**Comment by cinosh07** [comment]
> We did manage to fix the shared notes' Etherpad server recently to always set it's cookies to `SameSite=None` and `secure` (when behind a secured connection) because it always runs in an iframe [bigbluebutton/etherpad-lite@aafba50](https://github.com/bigbluebutton/etherpad-lite/commit/aafba502a4d19db8fd255c4c2fee9b3c8812df28).
> 
> Although might look like you could embed the whole bigbluebutton client in an iframe, that's not recommended. But in case you have no other alternative besides running it inside an iframe, try checking out how to configure that using nginx' `proxy_cookie_path`.

This solution worked for me. Thanks

**Comment by pedrobmarin** [comment]
> I think there is a different option, according to https://groups.google.com/g/bigbluebutton-dev/c/CP0trA-AKR0/m/UZEhBFtFAwAJ, BigBlueButton has an option: `allowRequestsWithoutSession`. That seems to eliminate the need to have a session and hence cookies.

This will only be correct if you disable your shared notes and closed captions. This may also affect the external video sharing but I'm not 100% on that one.

>     1. Is that an intended configuration option?

I would say this is a legacy/deprecated configuration left at bbb-web to, indeed, make the client be embedded in iframes BUT it only exists because of a webkit limitation (https://webkit.org/blog/7675/intelligent-tracking-prevention). BigBlueButton does not indicates in the docs (as far as I'm aware of) that the whole client can be embedded inside an iframe.

>     2. Are there any other side effects when using this option?

Besides the fact that this will not work for notes/captions/(and maybe external videos) you are correct assuming the security issue. I would say that's the main point of the session cookie

>     3. Would you assess the degree of security as high enough?

It really depends on what level of security you are offering your customers. Without session verification the url's session token  param might be the only door holding the client validation.

**Comment by somuelle-tmt** [comment]
@pedrobmarin Thanks for answering all these questions, I would like to comment one answer:

> BigBlueButton does not indicates in the docs (as far as I'm aware of) that the whole client can be embedded inside an iframe.
It does: https://docs.bigbluebutton.org/2.2/customize.html#external-parameters
There are even specific parameters for controlling the HTML5 clients from the parent page which we were using till last week where we had to build a workaround for Chrome.

**Comment by pedrobmarin** [comment]
> It does: https://docs.bigbluebutton.org/2.2/customize.html#external-parameters

@somuelle-tmt thanks for pointing that out. That looks like something that belong to the same former `allowRequestsWithoutSession` initiative. Unfortunately that's all I know about it too.

What I can say for sure is that there's no one validating and testing the client inside iframes right now. So take some extra cautions when doing so.

> No matter the client is in an iframe or not.

@cinosh07 I can't tell if this problem is related with the cookies configuration. I would say it isn't. Maybe check for any issue on 401's and breakout rooms.

**Comment by adeel-raza** [comment]
Is there any client-side solution for the above-mentioned issue on Chrome?
